### PR TITLE
Update gateway device page with the device configuration from scratch

### PR DESCRIPTION
**Pull request by Igoraamc**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This pull request updates the device page with a modal that the user can configure a new device from scratch manually.

Does this close any currently open issues?
------------------------------------------
Any relevant logs, error output, etc?
-------------------------------------
Any other comments?
-------------------
Where has this been tested?
---------------------------

**Operating System/Platform:** MACOS Catalina

**NodeJS Version:** v12.16.0

### CODE REVIEW

File: public/controllers.js [hunk 1]
```diff
@@ -436,3 +442,125 @@ appCtrls.controller('DevicesController', function DevicesController($scope, $int
...
+      $scope.serviceForm.thingd.dataItems[index].config.timeSec = config.timeSec;
+    }
+    clearDataItemForm();
+  }
+
+  $scope.next = function next() {
+    if ($scope.currentStep === 'CONFIG_DATA_ITEM') {
+      $scope.addDataItem($scope.form.dataItem, $scope.form.config);
+      $scope.currentStep = 'LIST_DATA_ITEMS';
+      $scope.title = 'List of all data items';
```
Comment (netoax) [inline public/controllers.js:533]:
```suggestion
      $scope.title = 'Data items';
```

File: public/controllers.js [hunk 2]
```diff
@@ -436,3 +442,125 @@ appCtrls.controller('DevicesController', function DevicesController($scope, $int
...
+      GatewayApi.createDevice($scope.serviceForm)
+        .then(function onSuccess() {
+          $uibModalInstance.dismiss('cancel');
+          return;
+      });
+    }
+
+    $scope.serviceForm.thingd = $scope.form.thingd;
+    $scope.currentStep = 'LIST_DATA_ITEMS';
+    $scope.title = 'List of all data items';
```
Comment (netoax) [inline public/controllers.js:545]:
```suggestion
    $scope.title = 'Data items';
```

Comment (netoax) [reply public/controllers.js:545]: I noticed these two lines are being repeated. Could they be moved to the top of function considering we don't change the values in the `else if` clause?

File: public/views/app.devices.html [hunk 3]
```diff
@@ -66,4 +66,122 @@ <h5>{{ device.name }}</h5>
...
+                <input type="number" class="form-control" ng-model="form.config.timeSec" placeholder="5">
+              </div>
+            </div>
+          </div>
+        </form>
+        <div class="data-item-list" ng-if="currentStep === 'LIST_DATA_ITEMS'">
+          <div class="col-sm-12 flex-center" ng-repeat="dataItem in serviceForm.thingd.dataItems">
+            <div class="row">
+              <div class="col-sm-1 data-item-index">
+                {{ $index + 1 }}
```
Comment (netoax) [inline public/views/app.devices.html:157]: I would remove this index from being shown in the page.

File: public/views/app.devices.html [hunk 4]
```diff
@@ -66,4 +66,122 @@ <h5>{{ device.name }}</h5>
...
+              </div>
+              <div class="col-sm-6 data-item">
```
Comment (netoax) [inline public/views/app.devices.html:159]: After removing the index, what do you think about increase the size of this div?

File: public/views/app.devices.html [hunk 5]
```diff
@@ -66,4 +66,122 @@ <h5>{{ device.name }}</h5>
...
+                {{ dataItem.schema.sensorName }}
+              </div>
+            </div>
+          </div>
+          <div class="col-sm-12 flex-center" ng-if="serviceForm.thingd.dataItems.length < 1">
+            <span>No data item configured.</span>
```
Comment (netoax) [inline public/views/app.devices.html:165]:
```suggestion
            <span>No data item configured</span>
```

File: public/views/app.devices.html [hunk 6]
```diff
@@ -66,4 +66,122 @@ <h5>{{ device.name }}</h5>
...
+          </div>
+          <div class="add-data-item">
+            <button class="add-data-item-button" api-click="changePage()">Add a new data item</button>
+          </div>
+        </div>
+      </div>
+      <div class="modal-footer">
+        <ul class="pager">
+          <li ng-if="currentStep !== 'LIST_DATA_ITEMS'" class="pull-right ng-class:{ disabled: deviceForm.$invalid }">
+            <a href api-click="next()" ng-disabled="deviceForm.$invalid">{{ currentStep === 'CREATE_DEVICE' ? 'Next' : 'Add data item' }}</a>
```
Comment (netoax) [inline public/views/app.devices.html:175]:
```suggestion
            <a href api-click="next()" ng-disabled="deviceForm.$invalid">{{ currentStep === 'CREATE_DEVICE' ? 'Next' : 'Save' }}</a>
```

File: public/controllers.js (hunk 1 above)
Comment (Igoraamc) [reply public/controllers.js:533]: Done!

File: public/views/app.devices.html (hunk 3 above)
Comment (Igoraamc) [reply public/views/app.devices.html:157]: Done!

File: public/views/app.devices.html (hunk 5 above)
Comment (Igoraamc) [reply public/views/app.devices.html:165]: Done!

File: public/views/app.devices.html (hunk 6 above)
Comment (Igoraamc) [reply public/views/app.devices.html:175]: Done!

File: public/views/app.devices.html (hunk 4 above)
Comment (Igoraamc) [reply public/views/app.devices.html:159]: Done!

File: public/controllers.js (hunk 2 above)
Comment (Igoraamc) [reply public/controllers.js]: Done!

File: public/styles/devices.less [hunk 7]
```diff
@@ -61,3 +65,85 @@
...
   text-align: right;
 }
+
+.add-device {
+  position: absolute;
+  right: 5%;
+  bottom: 0;
+}
+
+.add-device button {
```
Comment (netoax) [inline public/styles/devices.less:75]: This button change has some relation with the bug I reported to you (overlapping with device cards)? 

Independently, as this commit is getting bigger I advise you to break it into others.

File: public/views/app.devices.html [hunk 8]
```diff
@@ -66,4 +66,176 @@ <h5>{{ device.name }}</h5>
...
+                    <div class="col-sm-8 fields-space">
+                      <select class="form-control" ng-model="form.dataItem.schema.unit" required>
+                        <option ng-repeat="option in unitValues[form.dataItem.schema.typeID]"
+                        value="{{ option.value }}">{{ option.name }}</option>
+                      </select>
+                    </div>
+                    <label class="control-label col-sm-4" for="valueType">Value Type*:</label>
+                    <div class="col-sm-8 fields-space">
+                      <select class="form-control" ng-model="form.dataItem.schema.valueType" required>
+                        <option value="1" selected>INT</option>
```
Comment (netoax) [inline public/views/app.devices.html:152]: Couldn't they be lowercase as you did with type id's names?

File: public/views/app.devices.html [hunk 9]
```diff
@@ -66,4 +66,176 @@ <h5>{{ device.name }}</h5>
...
+              </div>
+            </div>
+          </div>
+          <div class="data-item-form" ng-if="currentStep === 'CONFIG_DATA_ITEM'">
+            <div class="form-group">
+              <div class="col-sm-12">
+                <div class="row">
+                  <div class="col-sm-6">
+                    <div class="col-sm-12 flex-center fields-title">
+                      <h3>KNoT</h3>
```
Comment (netoax) [inline public/views/app.devices.html:101]:
```suggestion
                      <h3>Schema</h3>
```

File: public/views/app.devices.html [hunk 10]
```diff
@@ -66,4 +66,176 @@ <h5>{{ device.name }}</h5>
...
+                      </select>
+                    </div>
+                    <label class="control-label col-sm-4" for="timeSec">Time Sec:</label>
+                    <div class="col-sm-8 fields-space">
+                      <input type="number" class="form-control" ng-model="form.config.timeSec" placeholder="5">
+                    </div>
+                  </div>
+                  <div class="col-sm-12">
+                    <div class="col-sm-12 flex-center fields-title">
+                      <h3>Modbus</h3>
```
Comment (netoax) [inline public/views/app.devices.html:188]: I think we can have a line above separating the sections in order to improve the visual (there's a lot of information tied together in the model). @Is4bela could help us on this.

File: public/views/app.devices.html (hunk 8 above)
Comment (Igoraamc) [reply public/views/app.devices.html:152]: Done!

File: public/views/app.devices.html (hunk 9 above)
Comment (Igoraamc) [reply public/views/app.devices.html:101]: Done!

File: public/views/app.devices.html (hunk 10 above)
Comment (Igoraamc) [reply public/views/app.devices.html:189]: Done!

File: public/controllers.js [hunk 11]
```diff
@@ -436,3 +441,125 @@ appCtrls.controller('DevicesController', function DevicesController($scope, $int
...
+      },
+      config: {}
+    };
+    $scope.form.config = {
+      lowerThreshold: undefined,
+      upperThreshold: undefined,
+      change: 'yes',
+      timeSec: undefined
+    };
+  }
```
Comment (netoax) [inline public/controllers.js]: Couldn't we reuse the `$scope.form.dataItem` (maybe creating an `emptyForm` there) defined above to avoid repetition?

Comment (Igoraamc) [reply public/controllers.js]: Done!

**Pull request merged**